File: problems/1456/1456.maximum-number-of-vowels-in-a-substring-of-given-length.py
#
# @lc app=leetcode id=1456 lang=python3
#
# [1456] Maximum Number of Vowels in a Substring of Given Length
#


# @lc code=start
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        vowels = "aeiou"
        curr_max = 0
        for i in range(k):
            if s[i] in vowels:
                curr_max += 1

        curr_cnt = curr_max
        for r in range(k, len(s)):
            if s[r - k] in vowels:
                curr_cnt -= 1
            if s[r] in vowels:
                curr_cnt += 1

            curr_max = max(curr_max, curr_cnt)

        return curr_max


# @lc code=end

File: problems/1456/1456.maximum-number-of-vowels-in-a-substring-of-given-length.go
/*
 * @lc app=leetcode id=1456 lang=golang
 *
 * [1456] Maximum Number of Vowels in a Substring of Given Length
 */

package solution

// @lc code=start
func maxVowels(s string, k int) int {
	curr_max := 0
	for i := 0; i < k; i++ {
		if isVowel(s[i]) {
			curr_max++
		}
	}

	curr_cnt := curr_max
	for r := k; r < len(s); r++ {
		if isVowel(s[r-k]) {
			curr_cnt--
		}
		if isVowel(s[r]) {
			curr_cnt++
		}

		curr_max = max(curr_cnt, curr_max)
	}

	return curr_max
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func isVowel(c byte) bool {
	s := string(c)
	return s == "a" || s == "e" || s == "i" || s == "o" || s == "u"
}

// @lc code=end
